docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,455 @@
+
+This solution can be generalized to other ClusterAddonProviders by defining a CRD that specifies an add-on to be installed and includes a cluster selector. The controller would opt-in any Cluster matching the selector for orchestration: install the add-on if it is not present, update it if it is present and configuration in the CRD changed, and uninstall the add-on if the Cluster is no longer selected.
+
+#### Install and lifecycle manage the Package Manager
+
+Some Package Manager might require in-cluster components, like e.g. the kapp-controller, however this is not the case for Helm starting from v3, so this is a no-op.
+In case this is required, the ClusterAddonProvider should take care of this task as well.
+In the case of Helm, we will need to consider that each version of Helm has official support for 4 versions of Kubernetes. Because a HelmChartProxy will require a particular implementation of the helm (and downstream Kubernetes) libraries, we will want to expose that version support matrix when we maintain versions of HelmChartProxy itself.
+
+#### Mapping HelmChartProxy to Helm charts
+
+The Helm artifact repository is huge, and it might happen there are different Helm charts for the same add-on; also in many organizations/products, users are maintaining their own internal Helm repository with certified Helm charts.
+
+In order to provide a solution to this problem, the HelmChartProxy CRD allows to define the link to the Helm chart we want to install. The link to the repository and chart is the canonical, unique source for the Helm chart to install.
+Mapping Kubernetes version to add-ons version
+Cluster add-ons by definition have a lifecycle strictly linked to the Cluster lifecycle, and the main consequences of this is the fact that there is a strict correlation between the Cluster’s Kubernetes version and the add-on version to be installed in a Cluster.
+
+In some cases this correlation is explicit, like e.g. for CPI following the “Versioning Policy for External Cloud Providers”; in other cases the compatibility matrix is defined case by case in documentation, like for “Supported version” in the Calico documentation.
+ 
+On top of that, in many Cluster API installations, the add-on version an operator would like to use is limited to a list of certified versions defined in each organization/product vendor.
+
+We can rely upon existing functionality and capabilities in Helm to enforce Kubernetes version-specific requirements in a chart, including whether or not to install anything at all. It's worth noting that other add-on tools might not provide this functionality out of the box.
+
+#### Generating package configurations 
+
+Each package manager has its own specific configuration format, like e.g. Helm’s values.yaml file; on top of that, usually each add-on package has its own set of specific keys/data structures defining configuration values, like e.g. aws-calico configuration.
```

Comment:
I think it's just a Helm chart since I found [this link](https://docs.aws.amazon.com/eks/latest/userguide/calico.html) when Googling it. But I believe the idea is just that the values.yaml file can have any number of nested fields.

Reply:
Ah makes sense. The example (without link) might be a bit confusing

Reply:
+1 on adding a link